### PR TITLE
Reduce OnStartLine params

### PR DESCRIPTION
**Pull request by benaadams**
Reduce params passed to OnStartLine 
Don't rescan to determine if contains dots

/cc @davidfowl

### CODE REVIEW

**Comment by benaadams** [comment]
Vector 16

dev
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.3760 us | 0.0238 us |   1.00 | 726,727.42 |     349 B |
          PlaintextAbsoluteUri | 1.4858 us | 0.0213 us |   1.08 | 673,039.94 |     381 B |
 PipelinedPlaintextTechEmpower | 1.1270 us | 0.0172 us |   0.82 | 887,293.56 |     348 B |
                    LiveAspNet | 2.8045 us | 0.0212 us |   2.04 | 356,571.07 |   1.07 kB |
           PipelinedLiveAspNet | 2.4464 us | 0.0139 us |   1.78 | 408,769.96 |   1.07 kB |
                       Unicode | 4.9510 us | 0.0272 us |   3.60 | 201,979.14 |   1.85 kB |
              UnicodePipelined | 4.7408 us | 0.1150 us |   3.45 | 210,935.86 |   1.86 kB |
```
pr
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.3529 us | 0.0145 us |   1.00 | 739,161.25 |     349 B |
          PlaintextAbsoluteUri | 1.4260 us | 0.0188 us |   1.05 | 701,262.78 |     381 B |
 PipelinedPlaintextTechEmpower | 1.0846 us | 0.0086 us |   0.80 | 921,979.29 |     348 B |
                    LiveAspNet | 2.7420 us | 0.0563 us |   2.03 | 364,698.40 |   1.07 kB |
           PipelinedLiveAspNet | 2.4512 us | 0.0539 us |   1.81 | 407,956.72 |   1.07 kB |
                       Unicode | 4.8557 us | 0.0757 us |   3.59 | 205,945.12 |   1.85 kB |
              UnicodePipelined | 4.5331 us | 0.0342 us |   3.35 | 220,599.55 |   1.86 kB |
```

Vector 32

dev
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.3254 us | 0.0189 us |   1.00 | 754,482.56 |     349 B |
          PlaintextAbsoluteUri | 1.4130 us | 0.0144 us |   1.07 | 707,717.53 |     381 B |
 PipelinedPlaintextTechEmpower | 1.0799 us | 0.0160 us |   0.81 | 925,976.20 |     348 B |
                    LiveAspNet | 2.5584 us | 0.0411 us |   1.93 | 390,862.50 |   1.07 kB |
           PipelinedLiveAspNet | 2.2534 us | 0.0316 us |   1.70 | 443,773.13 |   1.07 kB |
                       Unicode | 3.8528 us | 0.0555 us |   2.91 | 259,551.42 |   1.85 kB |
              UnicodePipelined | 3.7975 us | 0.0421 us |   2.87 | 263,328.93 |   1.86 kB |
```
pr
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.2537 us | 0.0127 us |   1.00 | 797,610.75 |     349 B |
          PlaintextAbsoluteUri | 1.3136 us | 0.0179 us |   1.05 | 761,241.22 |     381 B |
 PipelinedPlaintextTechEmpower | 1.0375 us | 0.0220 us |   0.83 | 963,849.65 |     348 B |
                    LiveAspNet | 2.4754 us | 0.0245 us |   1.97 | 403,970.87 |   1.07 kB |
           PipelinedLiveAspNet | 2.2808 us | 0.0417 us |   1.82 | 438,443.15 |   1.07 kB |
                       Unicode | 3.8196 us | 0.0726 us |   3.05 | 261,808.47 |   1.85 kB |
              UnicodePipelined | 3.7811 us | 0.0326 us |   3.02 | 264,476.07 |   1.86 kB |
```

**Comment by benaadams** [comment]
Vector 16

dev
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.3760 us | 0.0238 us |   1.00 | 726,727.42 |     349 B |
          PlaintextAbsoluteUri | 1.4858 us | 0.0213 us |   1.08 | 673,039.94 |     381 B |
 PipelinedPlaintextTechEmpower | 1.1270 us | 0.0172 us |   0.82 | 887,293.56 |     348 B |
                    LiveAspNet | 2.8045 us | 0.0212 us |   2.04 | 356,571.07 |   1.07 kB |
           PipelinedLiveAspNet | 2.4464 us | 0.0139 us |   1.78 | 408,769.96 |   1.07 kB |
                       Unicode | 4.9510 us | 0.0272 us |   3.60 | 201,979.14 |   1.85 kB |
              UnicodePipelined | 4.7408 us | 0.1150 us |   3.45 | 210,935.86 |   1.86 kB |
```
pr
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.3529 us | 0.0145 us |   1.00 | 739,161.25 |     349 B |
          PlaintextAbsoluteUri | 1.4260 us | 0.0188 us |   1.05 | 701,262.78 |     381 B |
 PipelinedPlaintextTechEmpower | 1.0846 us | 0.0086 us |   0.80 | 921,979.29 |     348 B |
                    LiveAspNet | 2.7420 us | 0.0563 us |   2.03 | 364,698.40 |   1.07 kB |
           PipelinedLiveAspNet | 2.4512 us | 0.0539 us |   1.81 | 407,956.72 |   1.07 kB |
                       Unicode | 4.8557 us | 0.0757 us |   3.59 | 205,945.12 |   1.85 kB |
              UnicodePipelined | 4.5331 us | 0.0342 us |   3.35 | 220,599.55 |   1.86 kB |
```

Vector 32

dev
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.3254 us | 0.0189 us |   1.00 | 754,482.56 |     349 B |
          PlaintextAbsoluteUri | 1.4130 us | 0.0144 us |   1.07 | 707,717.53 |     381 B |
 PipelinedPlaintextTechEmpower | 1.0799 us | 0.0160 us |   0.81 | 925,976.20 |     348 B |
                    LiveAspNet | 2.5584 us | 0.0411 us |   1.93 | 390,862.50 |   1.07 kB |
           PipelinedLiveAspNet | 2.2534 us | 0.0316 us |   1.70 | 443,773.13 |   1.07 kB |
                       Unicode | 3.8528 us | 0.0555 us |   2.91 | 259,551.42 |   1.85 kB |
              UnicodePipelined | 3.7975 us | 0.0421 us |   2.87 | 263,328.93 |   1.86 kB |
```
pr
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.2537 us | 0.0127 us |   1.00 | 797,610.75 |     349 B |
          PlaintextAbsoluteUri | 1.3136 us | 0.0179 us |   1.05 | 761,241.22 |     381 B |
 PipelinedPlaintextTechEmpower | 1.0375 us | 0.0220 us |   0.83 | 963,849.65 |     348 B |
                    LiveAspNet | 2.4754 us | 0.0245 us |   1.97 | 403,970.87 |   1.07 kB |
           PipelinedLiveAspNet | 2.2808 us | 0.0417 us |   1.82 | 438,443.15 |   1.07 kB |
                       Unicode | 3.8196 us | 0.0726 us |   3.05 | 261,808.47 |   1.85 kB |
              UnicodePipelined | 3.7811 us | 0.0326 us |   3.02 | 264,476.07 |   1.86 kB |
```

**Comment by benaadams** [comment]
Probably should be a struct rather than an enum

**Comment by benaadams** [comment]
Still works as struct

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpParsingDetails.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            set => _details = ((_details | 1) ^ 1) | (int)value;
+        }
+
+        public HttpMethod Method
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => (HttpMethod)((_details >> 1) & 0xF);
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            set => _details = ((_details | 0x1E) ^ 0x1E) | (int)value;
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpParsingDetails.cs]: Could do `_details = (_details & ~0x1E) | (int)value`

Which would come out as `_details = (_details & 0xFFFFFFE1) | (int)value`

**Comment by benaadams** [comment]
Made the struct more approachable

**Comment by benaadams** [comment]
Could change Unknown version to 0; then make it work with that also - though it should throw an exception

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpStartLineParseInfo.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        private const int PathContainsDots = 128;
+
+        private int _details;
+
+        public HttpStartLineParseInfo(HttpMethod method)
+        {
+            _details = ((int)method << 1);
+        }
+
+        public HttpVersion Version
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpStartLineParseInfo.cs]: nit: call it `HttpVersion`

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs [hunk 3]
```diff
@@ -107,7 +109,16 @@ public KestrelHttpParser(IKestrelTrace log)
...
                     }
 
-                    pathEncoded = true;
+                    parseInfo.IsPathEncoded = true;
+                }
+                else if (ch == ByteDot)
+                {
+                    if (pathStart == -1)
+                    {
+                        pathStart = offset;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: I think this is illegal after #1470 cc @natemcmaster

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Do you mean throw here?

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Lots of test failures if throw for it

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Two of the throws above can be removed though...

**Comment by davidfowl** [comment]
@benaadams needs a rebase...

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs (hunk 3 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Still fails; seems is allowed (or tested for authority form)

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpRequestLineParseInfo.cs [hunk 4]
```diff
@@ -7,11 +7,13 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Http
...
-        // Http10 = 0,
-        // Http11 = 1,
-        // HttpMethod: 2 - 32
-        private const int PathEncoded = 64;
-        private const int PathContainsDots = 128;
+        // ** Bit usage
+        // Http10 =                           0b0000_0000_0000_0000,
+        // Http11 =                           0b0000_0000_0000_0001,
+        // HttpMethod:                        0b0000_0000_0011_1110
+        private const int PathEncoded =       0b0000_0000_0100_0000;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpRequestLineParseInfo.cs]: lol

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpRequestLineParseInfo.cs]: Went too far, too fast...

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/HttpRequestLineParseInfo.cs]: Followed up in https://github.com/aspnet/KestrelHttpServer/pull/1479

**Comment by benaadams** [comment]
https://github.com/aspnet/KestrelHttpServer/pull/1479 is better

**Comment by Drawaes** [comment]
I really shouldn't have highlighted those binary consts to you!

**Pull request closed without merging**